geomstats/learning/frechet_mean.py

```diff
@@ -0,0 +1,348 @@
+"""Frechet mean."""
+
+import math
+import warnings
+
+from sklearn.base import BaseEstimator
+
+import geomstats.backend as gs
+from geomstats.geometry.euclidean import EuclideanMetric
+from geomstats.geometry.minkowski import MinkowskiMetric
+
+EPSILON = 1e-4
+
+
+def variance(points,
+             base_point,
+             metric,
+             weights=None,
+             point_type='vector'):
+    """Variance of (weighted) points wrt a base point.
+
+    Parameters
+    ----------
+    points: array-like, shape=[n_samples, dimension]
+
+    weights: array-like, shape=[n_samples, 1], optional
+    """
+    if point_type == 'vector':
+        points = gs.to_ndarray(points, to_ndim=2)
+        base_point = gs.to_ndarray(base_point, to_ndim=2)
+    if point_type == 'matrix':
+        points = gs.to_ndarray(points, to_ndim=3)
+        base_point = gs.to_ndarray(base_point, to_ndim=3)
+    n_points = gs.shape(points)[0]
+
+    if weights is None:
+        weights = gs.ones((n_points, 1))
+
+    weights = gs.array(weights)
+    weights = gs.to_ndarray(weights, to_ndim=2, axis=1)
+
+    sum_weights = gs.sum(weights)
+
+    var = 0.
+
+    sq_dists = metric.squared_dist(base_point, points)
+    var += gs.einsum('nk,nj->j', weights, sq_dists)
+
+    var = gs.array(var)
+    var /= sum_weights
+
+    var = gs.to_ndarray(var, to_ndim=1)
+    var = gs.to_ndarray(var, to_ndim=2, axis=1)
+    return var
+
+
+def linear_mean(points, weights=None):
+    """Compute the frechet mean.
+
+    The Frechet mean of (weighted) points computed with the
+    Euclidean metric is the weighted average of the points
+    in the Euclidean space.
+
+    Parameters
+    ----------
+    points: array-like, shape=[n_samples, dimension]
+
+    weights: array-like, shape=[n_samples, 1], optional
+
+    Returns
+    -------
+    mean: array-like, shape=[1, dimension]
+    """
+    if isinstance(points, list):
+        points = gs.vstack(points)
+    points = gs.to_ndarray(points, to_ndim=2)
+    n_points = gs.shape(points)[0]
+
+    if isinstance(weights, list):
+        weights = gs.vstack(weights)
+    elif weights is None:
+        weights = gs.ones((n_points, ))
+
+    weighted_points = gs.einsum('n,nj->nj', weights, points)
```

Comment:
use ellipsis for vectorized einsum?

Reply:
YESSSS! I just checked, and tensorflow 2.0 now supports ellipsis! This is gonna make the vectorization so much easier 🎉